File: core/hs/tests/gcbor/Sum.hs
{-# LANGUAGE BlockArguments #-}
{-# LANGUAGE QuasiQuotes #-}
{-# LANGUAGE TemplateHaskell #-}

module Sum (spec) where

import Common (mkTest)
import Data.Int
import qualified Data.Vector as V
import Data.Word
import Test.Hspec
import Webar.Codec.GCbor.TH
import Webar.Text.Normalized (NFText, nfTxt)

data SUnit
  = SuA
  | SuB
  | SuVar3
  deriving (Show, Eq)

deriveSumGCbor
  (simpleSumOptions (toSnakeCase 2) mkEncodedName [])
  ''SUnit

unit :: Spec
unit = describe "unit" do
  mkTest "a" SuA "var_a"
  mkTest "b" SuB "var_b"
  mkTest "var3" SuVar3 "var_var3"

data SUnary
  = StUv Bool
  | StUv2 (V.Vector Word32)
  deriving (Show, Eq)

deriveSumGCbor
  (simpleSumOptions (toSnakeCase 2) mkEncodedName [])
  ''SUnary

unary :: Spec
unary = describe "unary" do
  mkTest "uv" (StUv True) "var_uv"
  mkTest "uv2" (StUv2 (V.fromList [123, 456])) "var_uv2"

data SNormal
  = SnNv1 Word8 Int16
  | SnNv2 Bool NFText
  deriving (Show, Eq)

deriveSumGCbor
  (simpleSumOptions (toSnakeCase 2) mkEncodedName [])
  ''SNormal

normal :: Spec
normal = describe "normal" do
  mkTest "nv1" (SnNv1 123 (-3200)) "var_nv1"
  mkTest "nv2" (SnNv2 False [nfTxt|123|]) "var_nv2"

data SRecord
  = SrRV1
      { srA :: Word32,
        srK :: Int16,
        srAb :: Int8,
        srC :: Bool
      }
  | SrRv2
      { srZ :: Bool,
        srFull :: Bool,
        srNew :: Word32
      }
  deriving (Show, Eq)

deriveSumGCbor
  (simpleSumOptions (toSnakeCase 2) (toSnakeCase 2) [])
  ''SRecord

record :: Spec
record = describe "record" do
  mkTest "rv1" SrRV1 {srA = 10, srK = -128, srAb = -10, srC = False} "var_rv1"
  mkTest "rv2" SrRv2 {srZ = False, srFull = True, srNew = 0} "var_rv2"

data SMixed
  = MvUnit
  | MvU2
  | MvUnary Int8
  | MvNormal Bool Bool
  | MvRecord
      { mvsC :: Int8,
        mvsAc :: Bool,
        mvsAb :: NFText
      }
  deriving (Show, Eq)

deriveSumGCbor
  (simpleSumOptions (toSnakeCase 0) (toSnakeCase 3) [])
  ''SMixed

mixed :: Spec
mixed = describe "mixed" do
  mkTest "unit" MvUnit "var_mv_unit"
  mkTest "u2" MvU2 "var_mv_u2"
  mkTest "unary" (MvUnary (-10)) "var_mv_unary"
  mkTest "normal" (MvNormal False True) "var_mv_normal"
  mkTest "record" MvRecord {mvsC = -1, mvsAc = True, mvsAb = [nfTxt|a|]} "var_mv_record"

spec :: Spec
spec = describe "sum" do
  unit
  unary
  normal
  record
  mixed

File: core/rs/tests/gcbor/sum.rs
mod unit {
    use crate::test_success;

    #[derive(Debug, PartialEq, Eq, webar_core_macros::GCborCodec)]
    #[gcbor(rename_variants = "snake_case")]
    enum Unit {
        A,
        B,
        Var3,
    }

    #[test]
    fn a() {
        test_success(Unit::A, include_bytes!("./data/var_a.bin"))
    }
    #[test]
    fn b() {
        test_success(Unit::B, include_bytes!("./data/var_b.bin"))
    }
    #[test]
    fn var3() {
        test_success(Unit::Var3, include_bytes!("./data/var_var3.bin"))
    }
}

mod unary {
    use crate::test_success;

    #[derive(Debug, PartialEq, Eq, webar_core_macros::GCborCodec)]
    #[gcbor(rename_variants = "snake_case")]
    enum Unary {
        Uv(bool),
        Uv2(Vec<u32>),
    }

    #[test]
    fn uv() {
        test_success(Unary::Uv(true), include_bytes!("./data/var_uv.bin"))
    }
    #[test]
    fn uv2() {
        test_success(
            Unary::Uv2(Vec::from([123, 456])),
            include_bytes!("./data/var_uv2.bin"),
        )
    }
}

mod normal {
    use webar_core::{nf_str, text::normalized::NFString};

    use crate::test_success;

    #[derive(Debug, PartialEq, Eq, webar_core_macros::GCborCodec)]
    #[gcbor(rename_variants = "snake_case")]
    enum Normal {
        Nv1(u8, i16),
        Nv2(bool, NFString),
    }

    #[test]
    fn nv1() {
        test_success(
            Normal::Nv1(123, -3200),
            include_bytes!("./data/var_nv1.bin"),
        )
    }
    #[test]
    fn nv2() {
        test_success(
            Normal::Nv2(false, nf_str!("123").to_nf_string()),
            include_bytes!("./data/var_nv2.bin"),
        )
    }
}

mod record {
    use crate::test_success;

    #[derive(Debug, PartialEq, Eq, webar_core_macros::GCborCodec)]
    #[gcbor(rename_variants = "snake_case")]
    enum Record {
        Rv1 { a: u32, k: i16, ab: i8, c: bool },
        Rv2 { z: bool, full: bool, new: u32 },
    }

    #[test]
    fn rv1() {
        test_success(
            Record::Rv1 {
                a: 10,
                k: -128,
                ab: -10,
                c: false,
            },
            include_bytes!("./data/var_rv1.bin"),
        )
    }
    #[test]
    fn rv2() {
        test_success(
            Record::Rv2 {
                z: false,
                full: true,
                new: 0,
            },
            include_bytes!("./data/var_rv2.bin"),
        )
    }
}

mod mixed {
    use webar_core::{nf_str, text::normalized::NFString};

    use crate::test_success;

    #[derive(Debug, PartialEq, Eq, webar_core_macros::GCborCodec)]
    #[gcbor(rename_variants = "snake_case")]
    enum Mixed {
        MvUnit,
        MvU2,
        MvUnary(i8),
        MvNormal(bool, bool),
        MvRecord { c: i8, ac: bool, ab: NFString },
    }

    #[test]
    fn unit() {
        test_success(Mixed::MvUnit, include_bytes!("./data/var_mv_unit.bin"))
    }
    #[test]
    fn u2() {
        test_success(Mixed::MvU2, include_bytes!("./data/var_mv_u2.bin"))
    }
    #[test]
    fn unary() {
        test_success(
            Mixed::MvUnary(-10),
            include_bytes!("./data/var_mv_unary.bin"),
        )
    }
    #[test]
    fn normal() {
        test_success(
            Mixed::MvNormal(false, true),
            include_bytes!("./data/var_mv_normal.bin"),
        )
    }
    #[test]
    fn record() {
        test_success(
            Mixed::MvRecord {
                c: -1,
                ac: true,
                ab: nf_str!("a").to_nf_string(),
            },
            include_bytes!("./data/var_mv_record.bin"),
        )
    }
}
